hw/top_earlgrey/syn/chip_earlgrey_asic.sdc

```diff
@@ -50,35 +50,71 @@ set USB_CLK_PIN u_ast/clk_src_usb_o
 # target is 48MHz, overconstrain by 5%
 set USB_TCK_TARGET_PERIOD 20.8
 set USB_TCK_PERIOD [expr $USB_TCK_TARGET_PERIOD*$CLK_PERIOD_FACTOR]
-#set_ideal_network [get_pins ${USB_CLK_PIN}]
-
+# USB clock uncertainty needs to be within 2500ppm
+set USB_CLOCK_UNCERTAINTY [expr $USB_TCK_PERIOD * .0025]
 create_clock -name USB_CLK -period ${USB_TCK_PERIOD} [get_pins ${USB_CLK_PIN}]
-set_clock_uncertainty ${SETUP_CLOCK_UNCERTAINTY} [get_clocks USB_CLK]
+set_clock_uncertainty ${USB_CLOCK_UNCERTAINTY} [get_clocks USB_CLK]
 
-set USBDEV_IOMUX_PATH top_earlgrey/u_usbdev/i_usbdev_iomux/cdc_io_to_usb/u_sync_1/gen_*u_impl*
-set USBDEV_OUTREG_PATH top_earlgrey/u_usbdev/usbdev_impl/u_usb_fs_nb_pe/u_usb_fs_tx
-# This requires knowledge of actual pin names, which are different depending on
-# whether we run this with tech lizbs or not.
+# This requires knowledge of actual pin names, hence we only run this if we're compiling against
+# real libs (i.e., not GTECH mode).
 if {$FOUNDRY_ROOT != ""} {
-    set USB_N_PIN  gen_flops[2]*.u_size_only_reg/D
-    set USB_P_PIN  gen_flops[3]*.u_size_only_reg/D
-    set USB_PIN    gen_flops[1]*.u_size_only_reg/D
-    set USB_D_PIN  u_usb_d_flop/gen_*u_impl*/gen_flops[0]*.u_size_only_reg/Q
-    set USB_OE_PIN u_oe_flop/gen_*u_impl*/gen_flops[0]*.u_size_only_reg/Q
-} else {
-    set USB_N_PIN  d_i[2]
-    set USB_P_PIN  d_i[3]
-    set USB_PIN    d_i[1]
-    set USB_D_PIN  u_usb_d_flop/gen_*u_impl*/q_o[0]
-    set USB_OE_PIN u_oe_flop/gen_*u_impl*/q_o[0]
+  # generic constraints to make sure all reg <-> pad paths have a constraint.
+  # specific constraints to minimize skew are further below.
+  set FLOP_PATH gen_*u_impl*/gen_flops?0?*?u_size_only_reg
+  set_max_delay 5 -from [get_pins top_earlgrey/u_usbdev/usbdev_impl/u_usb_fs_nb_pe/u_usb_fs_tx/u_*_flop/${FLOP_PATH}/Q] \
```

Comment:
Generic comment: when using "-from {blah,blah}/Q" you might need to use "-from {blah,blah}/Q*" instead, because some flop cells have QN pins for inverted outputs that will be missed here by this form of constraint.

Reply:
That is a good point! 

In this particular case, the flops are instantiated using `prim_flop`, which happens to use flops that have only `Q` underneath.

However, to make these constraints somewhat more generic, I have added a `*` so that they would also apply to regs that have `QN` outputs.